// enzyme-adapter-preact-pure/build-cjs/src/StringRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const preact_1 = require("preact");
class StringRenderer {
    constructor(options) {
        this._options = options;
    }
    render(el, context) {
        if (this._options.renderToString) {
            return this._options.renderToString(el, context);
        }
        else {
            const tempContainer = document.createElement('div');
            (0, preact_1.render)(el, tempContainer);
            const html = tempContainer.innerHTML;
            (0, preact_1.render)((0, preact_1.h)('unmount-me', {}), tempContainer);
            return html;
        }
    }
    simulateError(nodeHierarchy, rootNode, error) {
        throw new Error('Static rendering does not support simulating errors');
    }
    simulateEvent(node, eventName, args) {
        throw new Error('Static rendering does not support simulating events');
    }
    unmount() {
        // No-op
    }
    getNode() {
        return null;
    }
    batchedUpdates(fn) {
        fn();
    }
}
exports.default = StringRenderer;
